controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -227,10 +227,30 @@ func (r *BareMetalHostReconciler) Reconcile(ctx context.Context, request ctrl.Re
 		return ctrl.Result{Requeue: true, RequeueAfter: provisionerNotReadyRetryDelay}, nil
 	}
 
+	// If the unmanaged annotation is set we remove any host from the
+	// provisioner and take no further action
+	if annotations != nil {
+		if _, ok := annotations[metal3v1alpha1.UnmanagedAnnotation]; ok {
+			provResult, err := prov.Delete()
```

Comment:
One doubt I have about implementing this task in the main reconcile loop _outside_ the state machine it's that the annotations could bypass completely the (de)provisioning limit (currently introduced by @s3rj1k in #807)

Reply:
@andfasano That is true, but note we're not doing any deprovisioning here, it only deletes the host from Ironic, which should be a much less expensive operation, and will just get retried if for any reason it fails?

Reply:
Note that the reason I did it like this is the feedback I got on https://github.com/metal3-io/metal3-docs/pull/168 indicated (IIUC) that the preference is to avoid adding any new state for this, and also to avoid adding any `$state->Unmanaged` transition (since it'll be hard to reliably implement any subsequent `Unmanaged->$state` transition without adding more "last state" data to the status.)

Doing it in the reconcile loop means we avoid touching the state machine at all, so it's much easier to resume management via the state machine in future if/when the annotation is removed.

Reply:
> @andfasano That is true, but note we're not doing any deprovisioning here, it only deletes the host from Ironic, which should be a much less expensive operation, and will just get retried if for any reason it fails?

Agree that in this case the overall scenario sounds less expensive, but since the deleting was part of it I raised this discussion point. Such kind of [issues](https://github.com/metal3-io/baremetal-operator/pull/807#issuecomment-790802043) start to appear when dealing simultaneously with an high number of BMH, so probably the real question probably is about the chances of ending up in a similar difficult scenario when using the unmanaged annotation.

> the preference is to avoid adding any new state for this

Right, sorry if I wasn't clear but I was referring to the [delay mechanism](https://github.com/metal3-io/baremetal-operator/pull/807/files#diff-eeed3703d04e4c23a7d7af8cd0b7931b6b7990f23d826c49bdbc31c5f0a50291R165) currently being part of the host state machine as a very early check (potentially the presence of the annotation could be an additional triggering condition in the delay check, and the task post-poned after it)

Reply:
> Such kind of issues start to appear when dealing simultaneously with an high number of BMH

Understood - in this case there is no change in the state of the physical host, so I suspect these issues are not relevant, but perhaps @s3rj1k can confirm

The main risk AFAICS is that someone tries to annotate $many BMH resources simultaneously, then it's perhaps possible that the `prov.Delete()` will fail temporarily if the ironic-api is overloaded - in that case we just retry on the next reconcile.

> I was referring to the delay mechanism currently being part of the host state machine as a very early check

Right, but as proposed we'd have to either change the BMH state to `metal3v1alpha1.StateDeleting` (which I want to avoid so we can return easily to the existing state if/when the annotation is removed), or we'd have to make the state machine care about annotations as you mention.

If folks feel strongly this case really requires the limit/delay I can look at adding it, but I'm not currently convinced the additional complexity is justified?

Reply:
@hardys Yea, I also see a potential overload for ironic database backend on burst.

I would prefer that all ironic operations fall into delay mechanism and also keep state of BMH.

IMO: Changing state of BMH to de-provisioning makes unmanaged annotation useless, one could just de-provision a host and remove it from cluster.

Reply:
@zaneb potentially yes, since the 'prov.Delete()' is an action that will occupy a slot at the end. So I agree in moving the function later in the state machine, and moreover trying to grant a slot before applying it (as a side note, the delay mechanism will also take care of the backoff)

Reply:
> @hardys Yea, I also see a potential overload for ironic database backend on burst.
> I would prefer that all ironic operations fall into delay mechanism and also keep state of BMH.

Ack, sounds like there are multiple reasons to move this into the state-machines so I'll take a look at that.

> IMO: Changing state of BMH to de-provisioning makes unmanaged annotation useless, one could just de-provision a host and remove it from cluster.

The motivation is to avoid deprovisioning (and removing the BMH) in a multi-tier deployment - there should never be a reason to apply this annotation to a host in the de-provisioning state, so we should probably make that impossible via the state-machine rework.

Reply:
Moved to `host_state_machine.go` in the latest version